helper/linux/drive.cpp

```diff
@@ -0,0 +1,153 @@
+
+typedef QHash<QString, QVariant> Properties;
+typedef QHash<QString, Properties> InterfacesAndProperties;
+typedef QHash<QDBusObjectPath, InterfacesAndProperties> DBusIntrospection;
+Q_DECLARE_METATYPE(Properties)
+Q_DECLARE_METATYPE(InterfacesAndProperties)
+Q_DECLARE_METATYPE(DBusIntrospection)
+
+Drive::Drive(const QString &identifier)
+    : err(stderr), fileDescriptor(QDBusUnixFileDescriptor(-1)), identifier(identifier),
+      device(std::move(std::unique_ptr<QDBusInterface>(new QDBusInterface("org.freedesktop.UDisks2", identifier, "org.freedesktop.UDisks2.Block", QDBusConnection::systemBus())))),
+      path(qvariant_cast<QDBusObjectPath>(device->property("Drive")).path()),
+      drive(std::move(std::unique_ptr<QDBusInterface>(new QDBusInterface("org.freedesktop.UDisks2", path, "org.freedesktop.UDisks2.Drive", QDBusConnection::systemBus())))) {
+}
+
+/**
+ * Open drive for writing.
+ */
+void Drive::open() {
+    if (getDescriptor() == -1) {
+        QDBusReply<QDBusUnixFileDescriptor> reply = device->callWithArgumentList(QDBus::Block, "OpenForBenchmark", { Properties{ { "writable", true } } });
+        fileDescriptor = reply.value();
+        if (!fileDescriptor.isValid()) {
+            throw std::runtime_error(reply.error().message().toStdString());
+            fileDescriptor = QDBusUnixFileDescriptor(-1);
+        }
+    }
+}
+
+/**
+ * Close drive for writing.
+ */
+void Drive::close() {
+    fileDescriptor = QDBusUnixFileDescriptor(-1);
+}
+
+/**
+ * Write buffer directly to drive.
+ */
+void Drive::write(const void *buffer, std::size_t size) {
+    int fd = getDescriptor();
+    if (static_cast<std::size_t>(::write(fd, buffer, size)) != size) {
+        throw std::runtime_error(QObject::tr("Destination drive is not writable").toStdString());
+    }
+}
+
+/**
+ * Grab file descriptor.
+ */
+int Drive::getDescriptor() {
+    return fileDescriptor.fileDescriptor();
+}
+
+/**
+ * Create a new dos label on the partition. This essentially wipes all
+ * existing information about partitions.
+ */
+void Drive::wipe() {
+    QDBusReply<void> formatReply = device->call("Format", "dos", Properties());
+    if (!formatReply.isValid() && formatReply.error().type() != QDBusError::NoReply) {
+        throw std::runtime_error(formatReply.error().message().toStdString());
+    }
+}
+
+/**
+ * Fill the rest of the drive with a primary partition that uses the fat
+ * filesystem.
+ */
+QPair<QString, qint64> Drive::addPartition(quint64 offset, const QString &label) {
+    QDBusInterface partitionTable("org.freedesktop.UDisks2", identifier, "org.freedesktop.UDisks2.PartitionTable", QDBusConnection::systemBus());
+    const quint64 proposedSize = device->property("Size").toULongLong() - offset;
+    QDBusReply<QDBusObjectPath> partitionReply = partitionTable.call("CreatePartition", offset, proposedSize, "", "", Properties());
+    if (!partitionReply.isValid()) {
+        throw std::runtime_error(partitionReply.error().message().toStdString());
```

Comment:
I figured out how to work around this when using libparted directly. Poorly I don't do this in this case.